File: android-app/app/src/main/java/com/feelbachelor/doompedia/ranking/RankingConfig.kt
package com.feelbachelor.doompedia.ranking

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class RankingConfig(
    val version: Int,
    val weights: RankingWeights,
    val personalization: PersonalizationConfig,
    val guardrails: GuardrailsConfig,
    val search: SearchConfig,
)

@Serializable
data class RankingWeights(
    val interest: Double,
    val novelty: Double,
    val diversity: Double,
    val quality: Double,
    @SerialName("repetitionPenalty")
    val repetitionPenalty: Double,
)

@Serializable
data class PersonalizationConfig(
    val defaultLevel: String,
    val levels: Map<String, Double>,
    val learningRates: Map<String, Double>,
    val dailyDriftCap: Double,
    val topicClamp: TopicClamp,
)

@Serializable
data class TopicClamp(
    val min: Double,
    val max: Double,
)

@Serializable
data class GuardrailsConfig(
    val explorationFloor: Double,
    val windowSize: Int,
    val maxSameTopicInWindow: Int,
    val minDistinctTopicsInWindow: Int,
    val cooldownCards: Int,
)

@Serializable
data class SearchConfig(
    val typoDistance: Int,
    val typoMinQueryLength: Int,
    val maxTypoCandidates: Int,
    val maxResults: Int,
)

File: ios-app/Doompedia/Ranking/RankingConfig.swift
import Foundation

struct RankingConfig: Codable {
    let version: Int
    let weights: RankingWeights
    let personalization: PersonalizationConfig
    let guardrails: GuardrailsConfig
    let search: SearchConfig
}

struct RankingWeights: Codable {
    let interest: Double
    let novelty: Double
    let diversity: Double
    let quality: Double
    let repetitionPenalty: Double
}

struct PersonalizationConfig: Codable {
    let defaultLevel: String
    let levels: [String: Double]
    let learningRates: [String: Double]
    let dailyDriftCap: Double
    let topicClamp: TopicClamp
}

struct TopicClamp: Codable {
    let min: Double
    let max: Double
}

struct GuardrailsConfig: Codable {
    let explorationFloor: Double
    let windowSize: Int
    let maxSameTopicInWindow: Int
    let minDistinctTopicsInWindow: Int
    let cooldownCards: Int
}

struct SearchConfig: Codable {
    let typoDistance: Int
    let typoMinQueryLength: Int
    let maxTypoCandidates: Int
    let maxResults: Int
}

enum RankingConfigLoader {
    static func load() throws -> RankingConfig {
        guard let url = Bundle.main.url(forResource: "ranking-config.v1", withExtension: "json") else {
            throw NSError(domain: "Doompedia", code: 1, userInfo: [NSLocalizedDescriptionKey: "Missing ranking config asset"])
        }

        let data = try Data(contentsOf: url)
        return try JSONDecoder().decode(RankingConfig.self, from: data)
    }
}
